From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Sun, 12 Sep 2010 17:25:19 +0000 (+0100)
Subject: drm/i915: Ensure that the crtcinfo is populated during mode_fixup()
X-Git-Url: http://git.kernel.org/?p=linux%2Fkernel%2Fgit%2Fickle%2Fdrm-intel.git;a=commitdiff_plain;h=532db7fe1fd75f20f3abf959419d160fb7850aff

drm/i915: Ensure that the crtcinfo is populated during mode_fixup()

This should fix the mysterious mode setting failures reported during
boot up and after resume, generally for i8xx class machines.

Bugzilla: https://bugzilla.kernel.org/show_bug.cgi?id=16478
Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
---

--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -2373,11 +2373,19 @@ static bool intel_crtc_mode_fixup(struct
 				  struct drm_display_mode *adjusted_mode)
 {
 	struct drm_device *dev = crtc->dev;
+
 	if (HAS_PCH_SPLIT(dev)) {
 		/* FDI link clock is fixed at 2.7G */
 		if (mode->clock * 3 > 27000 * 4)
 			return MODE_CLOCK_HIGH;
 	}
+
+	/* XXX some encoders set the crtcinfo, others don't.
+	 * Obviously we need some form of conflict resolution here...
+	 */
+	if (adjusted_mode->crtc_htotal == 0)
+		drm_mode_set_crtcinfo(adjusted_mode, 0);
+
 	return true;
 }
 
